In [2]:
!pip install --upgrade tables
!pip install eli5

     |████████████████████████████████| 4.3MB 4.9MB/s 
  Found existing installation: tables 3.4.4
    Uninstalling tables-3.4.4:
      Successfully uninstalled tables-3.4.4
     |████████████████████████████████| 112kB 5.0MB/s 


In [3]:
import pandas as pd
import numpy as np

from sklearn.dummy import DummyRegressor
from sklearn.tree import DecisionTreeRegressor

from sklearn.metrics import mean_absolute_error as mae
from sklearn.model_selection import cross_val_score

import eli5
from eli5.sklearn import PermutationImportance

/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.metrics.scorer module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.metrics. Anything that cannot be imported from sklearn.metrics is now part of the private API.
  warnings.warn(message, FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.feature_selection.base module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.feature_selection. Anything that cannot be imported from sklearn.feature_selection is now part of the private API.
  warnings.warn(message, FutureWarning)
Using TensorFlow backend.


In [21]:
cd "/content/drive/My Drive/Colab Notebooks/dw_matrix/matrix_two/dw_matrix_car/"

/content/drive/My Drive/Colab Notebooks/dw_matrix/matrix_two/dw_matrix_car


WCZYTYWANIE DANYCH


In [22]:
df = pd.read_hdf('data/car.h5')
df.shape

(106494, 155)

In [23]:
df.columns

Index(['breadcrumb', 'created_at', 'price_currency', 'price_details',
       'price_value', 'seller_address', 'seller_name', 'seller_type',
       'feature_czujniki-parkowania-przednie',
       'feature_poduszka-powietrzna-chroniąca-kolana',
       ...
       'param_pearl', 'param_stan', 'param_wersja', 'param_emisja-co2',
       'param_body-type', 'param_matowy', 'param_bezwypadkowy',
       'param_akryl-(niemetalizowany)', 'param_monthly-payment-value',
       'car_id'],
      dtype='object', length=155)

Dummy Model
first model without f.e.

In [24]:
df.select_dtypes(np.number).columns

Index(['price_value', 'car_id'], dtype='object')

In [25]:
feats = ['car_id']
X = df[feats].values
y = df['price_value'].values

model = DummyRegressor()
model.fit(X,y)
y_pred = model.predict(X)

mae(y, y_pred)



39465.934630440985

In [26]:
[x for x in df.columns if 'price' in x]

['price_currency', 'price_details', 'price_value']

In [27]:
df['price_currency'].value_counts()

PLN    106290
EUR       204
Name: price_currency, dtype: int64

Nomalize PLN and EUR 


In [28]:
df['price_currency'].value_counts(normalize=True) * 100

PLN    99.80844
EUR     0.19156
Name: price_currency, dtype: float64

PLN    99.80844
EUR     0.19156
Only 0,2% data in  database in col [price_currency] use EUR 
  

> *Indented block*



In [29]:
df['price_currency'].value_counts()

PLN    106290
EUR       204
Name: price_currency, dtype: int64

In [0]:
df = df[df[price_currency']].value_counts()

without normalize - del EUR because only 0,2% od data

In [30]:
df = df[ df['price_currency'] != 'EUR' ]
df.shape

(106290, 155)

In [31]:
df['price_currency'].value_counts()

PLN    106290
Name: price_currency, dtype: int64

Features

In [32]:
df.head()

breadcrumb  ... car_id
0    [Osobowe, Dacia, Logan, Dacia Logan II (2012-)]  ...      0
2  [Osobowe, Peugeot, 3008, Peugeot 3008 I (2009-...  ...      2
3  [Osobowe, Porsche, Cayenne, Porsche Cayenne II...  ...      3
4  [Osobowe, Peugeot, 308, Peugeot 308 T7 (2008-2...  ...      4
5  [Osobowe, Volkswagen, Golf, Volkswagen Golf VI...  ...      5

[5 rows x 155 columns]

In [33]:
for feat in df.columns:
  print(feat)

breadcrumb
created_at
price_currency
price_details
price_value
seller_address
seller_name
seller_type
feature_czujniki-parkowania-przednie
feature_poduszka-powietrzna-chroniąca-kolana
feature_kurtyny-powietrzne
feature_klimatyzacja-dwustrefowa
feature_światła-led
feature_czujnik-zmierzchu
feature_elektrycznie-ustawiane-lusterka
feature_asr-(kontrola-trakcji)
feature_poduszka-powietrzna-kierowcy
feature_cd
feature_elektryczne-szyby-przednie
feature_poduszka-powietrzna-pasażera
feature_system-start-stop
feature_światła-do-jazdy-dziennej
feature_komputer-pokładowy
feature_elektryczne-szyby-tylne
feature_klimatyzacja-manualna
feature_tapicerka-welurowa
feature_czujnik-deszczu
feature_światła-przeciwmgielne
feature_ogrzewanie-postojowe
feature_radio-niefabryczne
feature_regulowane-zawieszenie
feature_ogranicznik-prędkości
feature_zmieniarka-cd
feature_szyberdach
feature_isofix
feature_asystent-pasa-ruchu
feature_alufelgi
feature_bluetooth
feature_nawigacja-gps
feature_asystent-parkowania
fe

In [34]:
df['param_color'].factorize()  # change categorical data to numeric data - fuction factorize()

(array([-1, -1, -1, ..., -1, -1, -1]),
 Index(['Czarny', 'Srebrny', 'Złoty', 'Bordowy', 'Biały', 'Szary', 'Beżowy',
        'Zielony', 'Fioletowy', 'Niebieski', 'Inny kolor', 'Brązowy', 'Żółty'],
       dtype='object'))

Explore-  id arrays for id[0] - for 
 name of colors.

In [35]:
df['param_color'].factorize()[0]

array([-1, -1, -1, ..., -1, -1, -1])

* in this place - could add variable to pandas series and looking for sample()  or values()  

In [0]:
SUFFIX_CAT = '__cat'

for feat in df.columns:
  if isinstance(df[feat][0], list): continue
  
  factorized_values = df[feat].factorize()[0]
  if SUFFIX_CAT in feat:                
    # if value is exist not add new parameter - add value to sum of param
    df[feat] = factorized_values
  else:
    df[feat + SUFFIX_CAT] = factorized_values 
    # if value is not exist then add new cat param __a , __b etc

how many feats in values .
After that - not looking at price name param value because - need to predict model for price.  

In [40]:
cat_feats = [x for x in df.columns if SUFFIX_CAT in x ]
cat_feats
len(cat_feats)

154

In [42]:
cat_feats = [x for x in df.columns if SUFFIX_CAT in x ]
cat_feats = [x for x in cat_feats if 'price' not in x]    # minus 3 feats with name 'price'
len(cat_feats)

151

To predict without analize - all 151 feats

In [44]:
X = df[cat_feats].values
y = df['price_value'].values

model = DecisionTreeRegressor(max_depth=5)
scores = cross_val_score(model, X, y, cv=3, scoring='neg_mean_absolute_error')
# scores in minues value - minimalize to 0 
## better to predict
np.mean(scores)

-19566.588937368324

-19566.588937368324 much better score than before DUMMY_MODEL = 39465.934630440985

Below - 
next model m 
in pragmatic ways (not so fast)
# shows automaticly most important feats on heat_map table

In [45]:
m = DecisionTreeRegressor(max_depth=5)
m.fit(X,y)

imp = PermutationImportance(m,random_state=0).fit(X,y)
eli5.show_weights(imp, feature_names=cat_feats)